docs/enhancements/agent-upgrade.md

```diff
@@ -0,0 +1,196 @@
+problem:
+
+```
+$ oc get deployment assisted-service -n assisted-installer-stage -o json | \
+jq '.spec.template.spec.containers[].env[] | select(.name == "AGENT_DOCKER_IMAGE")'
+{
+  "name": "AGENT_DOCKER_IMAGE",
+  "value": "registry-proxy.engineering.redhat.com/rh-osbs/openshift4-assisted-installer-agent-rhel8:v1.0.0-141"
+}
+```
+
+In the integration and development environments we are using the `:latest` tag,
+which effectively disables this feature.
+
+The description of the `compatible-agent` validation will be the following:
+
+> The installation cannot start just yet because this host's agent is in the
+> process of being upgraded to a newer version, please wait, this could take
+> a few minutes.
+
+A new `upgrade-agent` step will be added to the service and to the agent.
+
+When the service receives a request for next steps from a host it will check if
+that host is in one of the `preparing-for-installation`, `installing`,
+`installing-in-progress`, `error` or `cancel` states. If it is then the service
+will return the next steps as usual. If it isn't then the service will retrieve
+from the database the result of the `compatible-agent` validation. If the
+validation passed then the service will return the next steps as usual,
+otherwise it will replace all the usual steps the `upgrade-agent` step.
+
+When the agent receives the `agent-upgrade` step it will do nothing and return
+an exit code indicating that the agent needs to be upgraded. When receiving that
+exit code the next step runner will also exit, with the same code. The main
+program will react to that exit code starting the next step runner with the new
+image.
+
+The `upgrade-agent` step is also (similar to register) something we must not
+break in order to allow agent upgrade.
+
+### Risks and Mitigations
+
+The main risk is introducing interference with the state machine that controls
+the hosts. To reduce the chances of that the upgrade will be performed only when
+the host is not in the process of installing, i.e. not in one of the
+`preparing-for-installation`, `installing`, `installing-in-progress`, `error` or
+`cancel` states.
+
+## Design Details
+
+None, see the implementation details above.
+
+### Open Questions
+
+- Do the _on premise_ installations use values of `AGENT_DOCKER_IMAGE` with the
+  `:latest` tag? If so that will need to be changed to use a unique version tag or
+  a SHA, otherwise the feature will be effectively disabled.
+
+- Do we want an different handling for the new `compatible-agent` validation in
+  the UI?
+
+### UI Impact
+
+The UI will at the very minimum need to display correctly the new
+`compatible-agent` validation. Without changes it will display the raw
+`compatible-agent` string as the name of the validation. We will probably want
+to change that to display a more user friendly name, like "Agent compatibility".
+
+We will probably want to change the way that the result of this validation is
```

Comment:
I think this is a significant ux change and we should talk with @nirfarkas

Reply:
Actually, we control the validation text in the backend.
This is the current message format, I think it's clear enough, but perhaps @nirfarkas can think of a better phrasing to reflect that this validation should resolve itself. 
```
The installation cannot start just yet because this host's agent is in the process of being upgraded to a newer version, please wait, this could take a few minutes.
``` 
In any case, I don't think this validation requires a different UX than any other validation.
The NTP sync validation is another validation that gets resolved automatically and presented to the user just the same.

Reply:
Without changes to the UI it will use the enum value without changes as the title of the validation, something like this:

**compatible-agent**: The installation ...

But we probably want to present it in a more human readable format. We already do it for other validations. Something like this:

**Agent compatibility**: The installation ...

That requires a small change in the UI.

We may also want to present it in a different category. Without changes it will be in the "Hardware" category. We could want to add a new "Software" category. That would require some extra changes in the UI.